# 02: Preprocessing

---

## Objective

Prepare the METABRIC dataset for the downstream survival-analysis and machine-learning workflows while avoiding unnecessary transformations and premature use of outcome information.

Two model-specific datasets are created:

- `df_survival.csv`: dataset prepared for Kaplan–Meier and Cox proportional hazards analyses.
- `df_ml.csv`: dataset reserved for predictive machine-learning analyses, where preprocessing and supervised feature selection will be fitted on training data only.

At this stage, preprocessing is intentionally limited to transformations that are common to the downstream analyses, including survival-target recoding, variable exclusions, mutation binarization, and exploratory low-variance genomic filtering.

Missing-value imputation, categorical encoding, scaling, and supervised feature selection are deferred to the corresponding modeling notebooks.

---

## Notebook Outline

0. Setup & Loading
1. Target Variable Recoding
2. Modeling Dataset Separation & Variable Exclusions
3. Survival Variable Blocks Redefinition
4. Mutation Binarization
5. Model-Specific Preprocessing Strategy
6. Genomic Feature Filtering
7. Modeling Datasets Export
8. Preprocessing Summary

## 0. Setup & Loading

We start from the raw dataset to ensure full traceability. All preprocessing steps will be applied to a working copy, while the original data remain unchanged.

A copy, `df_raw`, is retained to allow comparisons between the original and preprocessed data when needed.


In [1]:
import os
import json

import pandas as pd

from sklearn.feature_selection import VarianceThreshold

DATA_PATH = "../data/METABRIC_RNA_Mutation.csv"

df_raw = pd.read_csv(DATA_PATH, low_memory=False)
df = df_raw.copy()

print(f"{df.shape[0]} patients, {df.shape[1]} variables")

1904 patients, 693 variables


## 1. Target Variable Recoding

In METABRIC, `overall_survival` is encoded inversely to the event convention used for survival analysis:

| Original Value | Meaning | Event Encoding |
|---|---|---|
| `overall_survival = 0` | Death observed | `event = 1` |
| `overall_survival = 1` | Alive at last follow-up (censored) | `event = 0` |

We therefore create:

`event = 1 - overall_survival`

The `overall_survival_months` variable is kept unchanged and represents the observed time from diagnosis to either death or last follow-up.

Together, these two variables define the survival outcome:
- **duration:** `overall_survival_months`
- **event indicator:** `event`

In [2]:
df["event"] = 1 - df["overall_survival"]

print("=" * 30)
print("EVENT RECODING CHECK")
print("=" * 30)

event_check = (
    df.groupby("event")["overall_survival_months"]
    .agg(["count", "median"])
    .round(1)
    .reset_index()
    .rename(columns={
        "count": "n",
        "median": "Median observed time (months)"
    })
)

display(event_check)

EVENT RECODING CHECK


,event,n,Median observed time (months)
0,0,801,158.0
1,1,1103,85.9


> **Recoding confirmed**:
> - `event = 0` (alive/censored): median observed time ~158 months
> - `event = 1` (death): median observed time ~86 months
>
> The pattern is consistent with the EDA. `event` will be used as the event indicator in all survival models, with `overall_survival_months` as the duration variable.

## 2. Exclusion of Variables Not Retained for Modeling

Based on the EDA and the planned modeling strategy, several variables are excluded before further preprocessing.

Variables may be excluded because of substantial missingness, redundancy, limited relevance to the planned models, identifier status, or potential target leakage.

### Modeling Dataset Separation

The project includes two distinct modeling objectives:

- **Survival analysis (`df_survival`)**: Kaplan–Meier and Cox proportional hazards models, using observed survival time and the event indicator.
- **Machine learning (`df_ml`)**: fixed-time survival prediction, for which the target and preprocessing strategy will be defined separately.

Separate datasets are therefore created before model-specific exclusions. This prevents decisions made for the Cox model from unnecessarily removing variables that may still be evaluated in the ML pipeline.

In [3]:
df_survival = df.copy()
df_ml = df.copy()

print("Initial datasets:")
print(f"Survival : {df_survival.shape}")
print(f"ML       : {df_ml.shape}")

Initial datasets:
Survival : (1904, 694)
ML       : (1904, 694)


In [4]:
# ============================================================
# VARIABLES EXCLUDED FROM ALL MODELING
# ============================================================

common_drop = [
    "patient_id",                   # identifier
    "overall_survival",             # replaced by event
    "death_from_cancer",            # target leakage
    "er_status_measured_by_ihc",    # redundant with er_status
    "her2_status_measured_by_snp6", # redundant with her2_status
    "oncotree_code",                # overlapping tumor-type information
]

common_drop = [c for c in common_drop if c in df.columns]

df_survival = df_survival.drop(columns=common_drop)
df_ml = df_ml.drop(columns=common_drop)
print(f"{len(common_drop)} columns dropped")
print(f"Shape: {df.shape}")

6 columns dropped
Shape: (1904, 694)


In [5]:
# ============================================================
# SURVIVAL-SPECIFIC EXCLUSIONS
# ============================================================

survival_drop = [
    "tumor_stage",               # 26.3% missing; excluded from primary Cox strategy
    "3-gene_classifier_subtype", # 10.7% missing; not retained in primary model
    "primary_tumor_laterality",  # limited expected prognostic contribution
]

survival_drop = [
    c for c in survival_drop
    if c in df_survival.columns
]

df_survival = df_survival.drop(columns=survival_drop)
print(f"{len(survival_drop)} columns dropped")
print(f"Shape: {df_survival.shape}")

3 columns dropped
Shape: (1904, 685)


In [6]:
# ============================================================
# ML DATASET
# ============================================================

# tumor_stage, 3-gene classifier subtype and laterality are
# retained at this stage. Their usefulness and missing-data
# handling will be evaluated within the ML preprocessing pipeline.

print("=" * 40)
print("MODELING DATASETS")
print("=" * 40)

print(f"Survival dataset : {df_survival.shape}")
print(f"ML dataset       : {df_ml.shape}")

print("\nExcluded from both:")
for col in common_drop:
    print(f"  - {col}")

print("\nAdditional exclusions from survival dataset:")
for col in survival_drop:
    print(f"  - {col}")

MODELING DATASETS
Survival dataset : (1904, 685)
ML dataset       : (1904, 688)

Excluded from both:
  - patient_id
  - overall_survival
  - death_from_cancer
  - er_status_measured_by_ihc
  - her2_status_measured_by_snp6
  - oncotree_code

Additional exclusions from survival dataset:
  - tumor_stage
  - 3-gene_classifier_subtype
  - primary_tumor_laterality


> **Preprocessing decision**
>
> Six variables are excluded from both modeling datasets because they represent identifiers, redundant information, or potential target leakage.
>
> Three additional variables (`tumor_stage`, `3-gene_classifier_subtype`, and `primary_tumor_laterality`) are excluded from the primary survival-modeling dataset but retained in the ML dataset at this stage. Their potential predictive value and missing-data handling will be evaluated separately during the ML preprocessing pipeline.
>
> As a result, `df_survival` contains 685 variables and `df_ml` contains 688 variables. Both datasets retain all 1,904 patients at this stage.

### Rationale for Key Exclusions

> **`death_from_cancer` — target leakage**
>
> This variable contains information about the patient's vital status and cause of death. Because this information is determined during follow-up and is directly related to the survival outcome, it is excluded from both modeling datasets to prevent target leakage.
>
> **`tumor_stage` — substantial missingness**
>
> `tumor_stage` is missing for 26.3% of patients. Missingness was associated with observed survival status in the EDA (28.2% among patients with an observed death vs. 23.7% among censored patients; χ² = 4.57, p = 0.0326), providing evidence against the independence expected under a strict MCAR mechanism.
>
> However, the available data do not allow MAR and MNAR mechanisms to be distinguished. Rather than applying simple imputation to a variable with substantial missingness, `tumor_stage` is excluded from the primary survival-modeling strategy. It is retained in the ML dataset at this stage so that its potential predictive value and an appropriate missing-data strategy can be evaluated separately.
>
> NPI is retained in the survival dataset as a composite clinical prognostic variable because it incorporates information from tumor size, histological grade, and lymph-node involvement. NPI and tumor stage capture partially overlapping clinical information but are not considered interchangeable.

## 3. Survival Variable Blocks Redefinition

Following the model-specific exclusions, variable blocks are redefined for `df_survival`.

These blocks will determine how each type of variable is handled during preprocessing for survival analyses.

| Block | Planned Processing |
|---|---|
| `target_cols` | Survival duration and event indicator; excluded from feature preprocessing |
| `treatment_cols` | Already binary (0/1); no encoding required |
| `biomarker_cols` | Binary categorical variables; encoding required |
| `mut_cols` | Mutation annotations; converted to mutation presence/absence (0/1) |
| `expr_cols` | Continuous gene-expression features |
| `clinical_cols` | Processed according to variable type and missingness |

In [7]:
print("=" * 40)
print("SURVIVAL VARIABLE BLOCKS")
print("=" * 40)

target_cols = [
    "overall_survival_months",
    "event"
]

treatment_cols = [
    "chemotherapy",
    "hormone_therapy",
    "radio_therapy"
]

biomarker_cols = [
    "er_status",
    "pr_status",
    "her2_status"
]

mut_cols = [
    c for c in df_survival.columns
    if c.endswith("_mut")
]

clinical_cols = [
    "age_at_diagnosis",
    "type_of_breast_surgery",
    "cancer_type",
    "cancer_type_detailed",
    "cellularity",
    "cohort",
    "neoplasm_histologic_grade",
    "tumor_other_histologic_subtype",
    "inferred_menopausal_state",
    "integrative_cluster",
    "lymph_nodes_examined_positive",
    "mutation_count",
    "nottingham_prognostic_index",
    "pam50_+_claudin-low_subtype",
    "tumor_size"
]

already_assigned = (
    target_cols
    + treatment_cols
    + biomarker_cols
    + mut_cols
    + clinical_cols
)

expr_cols = [
    c for c in df_survival.columns
    if c not in already_assigned
]

groups = {
    "Target": target_cols,
    "Treatments": treatment_cols,
    "Biomarkers": biomarker_cols,
    "Mutations": mut_cols,
    "Clinical": clinical_cols,
    "Expression": expr_cols
}

summary = pd.DataFrame({
    "Variables": {name: len(cols) for name, cols in groups.items()}
})

summary.loc["Total"] = summary["Variables"].sum()

display(summary)

total = summary.loc["Total", "Variables"]

print(f"Check: {total} / {df_survival.shape[1]}")

SURVIVAL VARIABLE BLOCKS


,Variables
Target,2
Treatments,3
Biomarkers,3
Mutations,173
Clinical,15
Expression,489
Total,685


Check: 685 / 685


> **Structure check**
>
> The redefined variable blocks account for all 685 columns in `df_survival`, including 489 gene-expression features and 173 mutation features.
>
> Each variable is assigned to exactly one block, ensuring that subsequent preprocessing steps can be applied according to variable type without modifying the survival outcomes.

## 4. Mutation Binarization

The 173 `_mut` columns contain specific variant annotations (e.g. `H1047R`, `E542K`) rather than simple binary indicators.

For this project, mutation features are represented at the **gene level**:

- `0` = no recorded mutation
- `1` = at least one recorded mutation

This reduces the dimensionality and avoids treating individual variant names as categorical levels. Variant-specific effects are beyond the scope of the current analysis.

In [8]:
print("=" * 40)
print("MUTATION BINARIZATION")
print("=" * 40)

# Check missing values before binarization
n_missing_mut = df_survival[mut_cols].isna().sum().sum()
print(f"Missing values across mutation columns: {n_missing_mut}")

# Binarize mutation annotations in both modeling datasets
for dataset in [df_survival, df_ml]:
    dataset[mut_cols] = (
        dataset[mut_cols]
        .ne("0")
        .astype(int)
    )

print(f"{len(mut_cols)} mutation columns binarized")

# Display 5 examples
example_mutations = [
    "pik3ca_mut",
    "tp53_mut",
    "muc16_mut",
    "kmt2c_mut",
    "map3k1_mut"
]

print("\nExamples:")
for col in example_mutations:
    counts = df_survival[col].value_counts().sort_index()
    print(
        f"{col:<15} "
        f"Absent (0): {counts.get(0, 0):4d} | "
        f"Present (1): {counts.get(1, 0):4d}"
    )

MUTATION BINARIZATION
Missing values across mutation columns: 0
173 mutation columns binarized

Examples:
pik3ca_mut      Absent (0): 1109 | Present (1):  795
tp53_mut        Absent (0): 1245 | Present (1):  659
muc16_mut       Absent (0): 1578 | Present (1):  326
kmt2c_mut       Absent (0): 1670 | Present (1):  234
map3k1_mut      Absent (0): 1706 | Present (1):  198


> **Binarization decision**
>
> Mutation annotations are converted to gene-level mutation indicators. A value of `1` indicates that at least one mutation is recorded for the gene, while `0` indicates no recorded mutation.
>
> This representation intentionally discards variant-specific information (e.g. `H1047R` vs. `E545K`). The objective of this project is to evaluate gene-level mutation status rather than the biological or clinical significance of individual variants.

## 5. Model-Specific Preprocessing Strategy

The project includes survival analyses and predictive machine-learning models with different preprocessing requirements. Therefore, no global imputation, scaling, or categorical encoding is applied at this stage.

Applying a single preprocessing pipeline to the entire dataset would unnecessarily transform variables that may not be used by a given model and could introduce data leakage in the predictive modeling workflow.

Instead, preprocessing will be performed within each downstream analysis:

- **Kaplan–Meier analysis:** uses the original grouping variables together with `overall_survival_months` and `event`. No scaling is required.
- **Cox proportional hazards models:** preprocessing will be restricted to the predictors selected for each Cox model. Missing-value handling and categorical encoding will therefore be performed only when required.
- **Machine-learning models:** imputation, encoding, scaling when appropriate, and feature selection will be fitted on the training data only and subsequently applied to the test data.

The mutation features are the only genomic variables transformed at this stage, as their gene-level presence/absence representation is common to the planned downstream analyses.

In [9]:
print("=" * 40)
print("FINAL PREPROCESSING CHECK")
print("=" * 40)

print(f"Survival dataset : {df_survival.shape}")
print(f"ML dataset       : {df_ml.shape}")

print("\nMissing values:")
print(f"df_survival : {df_survival.isna().sum().sum()}")
print(f"df_ml       : {df_ml.isna().sum().sum()}")

print("\nMutation encoding check:")
mutation_values = sorted(
    df_survival[mut_cols]
    .stack()
    .unique()
)

print(f"Unique values in _mut columns: {mutation_values}")

print("\nSurvival outcome:")
print(f"Duration column : overall_survival_months")
print(f"Event column    : event")
print(df_survival["event"].value_counts().sort_index())

FINAL PREPROCESSING CHECK
Survival dataset : (1904, 685)
ML dataset       : (1904, 688)

Missing values:
df_survival : 243
df_ml       : 1054

Mutation encoding check:
Unique values in _mut columns: [np.int64(0), np.int64(1)]

Survival outcome:
Duration column : overall_survival_months
Event column    : event
event
0     801
1    1103
Name: count, dtype: int64


> **Preprocessing status**
>
> Two model-specific datasets have been prepared while preserving all 1,904 patients:
>
> - `df_survival`: dataset prepared for survival analyses
> - `df_ml`: dataset reserved for the predictive modeling workflow
>
> The survival outcome has been standardized as `overall_survival_months` (duration) and `event` (1 = death observed, 0 = censored).
>
> Mutation annotations have been converted to binary gene-level indicators (`0` = no recorded mutation, `1` = mutation present).
>
> Missing values and categorical variables have intentionally not been processed globally. Their treatment will be defined within each downstream modeling workflow to avoid unnecessary transformations and, for predictive models, prevent information leakage.

## 6. Genomic Feature Filtering

The dataset contains a high-dimensional genomic block composed of gene-expression features and binary gene-level mutation indicators.

Feature selection must be handled carefully because the project contains censored survival data and predictive machine-learning analyses.

At this preprocessing stage, only **unsupervised filtering** is performed:

- gene-expression features are retained without target-based selection;
- mutation features with extremely low variability are removed from the candidate genomic pool.

No genomic feature is selected here according to `overall_survival_months` or `event`. Target-based feature selection will be performed within the appropriate downstream modeling workflow.

This avoids using outcome information prematurely and preserves the distinction between survival analysis and predictive modeling.

In [10]:
from sklearn.feature_selection import VarianceThreshold

print("=" * 40)
print("GENOMIC FEATURE FILTERING")
print("=" * 40)

print(f"Expression features : {len(expr_cols)}")
print(f"Mutation features   : {len(mut_cols)}")
print(f"Total genomic       : {len(expr_cols) + len(mut_cols)}")

# Mutation columns have already been binarized to 0/1
mutation_data = df_survival[mut_cols]

# Remove very low-variance mutation features
var_selector = VarianceThreshold(threshold=0.01)
var_selector.fit(mutation_data)

mut_cols_filtered = mutation_data.columns[
    var_selector.get_support()
].tolist()

removed_mutations = [
    col for col in mut_cols
    if col not in mut_cols_filtered
]

print("\nAfter mutation variance filtering:")
print(f"Mutations retained : {len(mut_cols_filtered)}")
print(f"Mutations removed  : {len(removed_mutations)}")

# Candidate genomic feature pool
genomic_cols_filtered = expr_cols + mut_cols_filtered

print(f"\nGenomic candidate pool: {len(genomic_cols_filtered)} features")

GENOMIC FEATURE FILTERING
Expression features : 489
Mutation features   : 173
Total genomic       : 662

After mutation variance filtering:
Mutations retained : 115
Mutations removed  : 58

Genomic candidate pool: 604 features


> **Genomic filtering results**
>
> The initial genomic block contained 662 features: 489 gene-expression features and 173 binary mutation features.
>
> Low-variance filtering was applied only to the mutation features. Using a `VarianceThreshold` of 0.01, 58 rare or near-constant mutation features were removed, leaving 115 mutation features.
>
> Gene-expression features were not filtered at this stage because no outcome-independent selection criterion was applied to them.
>
> The resulting genomic candidate pool therefore contains **604 features**:
> - 489 gene-expression features
> - 115 mutation features
>
> No outcome-based genomic feature selection is performed in this preprocessing notebook. Supervised feature selection will be handled separately within the downstream modeling workflows.

## 7. Modeling Datasets Export

Two modeling datasets are exported for the downstream analyses:

- **`df_survival`** contains the variables retained for survival analyses. No outcome-based genomic feature selection has been applied at this stage.
- **`df_ml`** preserves the candidate predictors required for predictive modeling. Imputation, encoding, scaling when appropriate, low-variance filtering, and supervised feature selection will be fitted on the training data within the ML workflow.

The genomic candidate pool identified during preprocessing contains 489 gene-expression features and 115 mutation features after exploratory low-variance filtering. However, the ML dataset retains the original mutation candidate set so that variance filtering can be learned from training data only.

This separation prevents preprocessing decisions from one modeling objective from constraining the other.

In [11]:
print("=" * 40)
print("FINAL DATASET CHECK")
print("=" * 40)

summary_export = pd.DataFrame({
    "Patients": [
        df_survival.shape[0],
        df_ml.shape[0]
    ],
    "Variables": [
        df_survival.shape[1],
        df_ml.shape[1]
    ]
}, index=["df_survival", "df_ml"])

display(summary_export)

print(f"Survival dataset : {df_survival.shape}")
print(f"ML dataset       : {df_ml.shape}")

print("\nSurvival outcome check:")
print(df_survival["event"].value_counts().sort_index())

print("\nGenomic candidate pool:")
print(f"Expression features           : {len(expr_cols)}")
print(f"Original mutation features    : {len(mut_cols)}")
print(f"Variance-filtered mutations   : {len(mut_cols_filtered)}")
print(f"Exploratory genomic pool      : {len(genomic_cols_filtered)}")

FINAL DATASET CHECK


,Patients,Variables
df_survival,1904,685
df_ml,1904,688


Survival dataset : (1904, 685)
ML dataset       : (1904, 688)

Survival outcome check:
event
0     801
1    1103
Name: count, dtype: int64

Genomic candidate pool:
Expression features           : 489
Original mutation features    : 173
Variance-filtered mutations   : 115
Exploratory genomic pool      : 604


In [12]:
OUTPUT_DIR = "../data/processed"
os.makedirs(OUTPUT_DIR, exist_ok=True)

df_survival.to_csv(
    f"{OUTPUT_DIR}/df_survival.csv",
    index=False
)

df_ml.to_csv(
    f"{OUTPUT_DIR}/df_ml.csv",
    index=False
)

In [13]:
feature_groups = {
    "expression": expr_cols,
    "mutations": mut_cols,
    "mutations_variance_filtered": mut_cols_filtered,
    "genomic_variance_filtered": genomic_cols_filtered,
    "clinical": clinical_cols,
    "treatments": treatment_cols,
    "biomarkers": biomarker_cols,
    "targets": target_cols
}

with open(
    f"{OUTPUT_DIR}/feature_groups.json",
    "w"
) as fh:
    json.dump(feature_groups, fh, indent=2)

print("\nFiles exported:")

for filename in [
    "df_survival.csv",
    "df_ml.csv",
    "feature_groups.json"
]:
    path = f"{OUTPUT_DIR}/{filename}"
    size = os.path.getsize(path) / 1024
    print(f"  {filename:<25} {size:>8.1f} KB")


Files exported:
  df_survival.csv             7875.5 KB
  df_ml.csv                   7921.0 KB
  feature_groups.json           21.3 KB


## Preprocessing Summary

| Step | Action | Result |
|---|---|---|
| Survival outcome | `event = 1 - overall_survival` | `1` = death observed, `0` = censored |
| Dataset separation | Created survival and ML datasets | Model-specific preprocessing can be performed independently |
| Variable exclusions | Removed identifiers, leakage variables, redundant variables, and survival-specific exclusions | Reduced unnecessary or potentially problematic predictors |
| Mutation representation | 173 `_mut` columns converted to binary indicators | `0` = no recorded mutation, `1` = mutation present |
| Genomic filtering assessment | Low-variance filtering evaluated on mutation features | 173 → 115 mutation candidates; 604 genomic candidates including expression |
| Missing data | No global imputation applied | Missingness will be handled according to each downstream model |
| Encoding | No global categorical encoding applied | Encoding will be performed only when required |
| Scaling | No global scaling applied | Scaling will be model-specific when appropriate |
| Feature selection | No outcome-based selection applied | Prevents premature use of survival outcomes and predictive leakage |
| Export | `df_survival.csv`, `df_ml.csv`, `feature_groups.json` | Ready for downstream modeling workflows |

> **Important modeling principle**
>
> Preprocessing is intentionally model-specific. Kaplan–Meier analysis can use the original grouping variables, Cox models will preprocess only their selected predictors, and the predictive ML pipeline will learn imputation, encoding, filtering, scaling when appropriate, and supervised feature selection from training data only.

**Next step:** `03_kaplan_meier.ipynb`